pybop/parameterisation.py

```diff
@@ -41,15 +63,98 @@ def __init__(self, model, observations, fit_parameters, x0=None, options=None):
+        # set input parameters in parameter set from fitting parameters
+        for i in self.fit_parameters:
+            self.parameter_set[i] = "[input]"
+
+        self._unprocessed_parameter_set = self.parameter_set
+        self.fit_dict = {
+            p: self.fit_parameters[p].prior.mean for p in self.fit_parameters
+        }
+
+        # Set t_eval
+        self.time_data = self.parameter_set["Current function [A]"].x[0]
+
+        self.build_model(check_model=check_model, init_soc=init_soc)
+
+    def build_model(self, check_model=True, init_soc=None):
+        """
+        Build the model (if not built already).
+        """
+        if init_soc is not None:
+            self.set_init_soc(init_soc)  # define this function
+
+        if self._built_model:
+            return
+        elif self.model.is_discretised:
+            self.model._model_with_set_params = self.model
+            self.model._built_model = self.model
+        else:
+            self.set_params()
+            self._mesh = pybamm.Mesh(self._geometry, self._submesh_types, self._var_pts)
+            self._disc = pybamm.Discretisation(self._mesh, self._spatial_methods)
+            self._built_model = self._disc.process_model(
+                self._model_with_set_params, inplace=False, check_model=check_model
+            )
+
+            # Clear solver
+            self.solver._model_set_up = {}
+
+    def set_init_soc(self, init_soc):
+        """
+        Set the initial state of charge.
+        """
+        if self._built_initial_soc != init_soc:
+            # reset
+            self._model_with_set_params = None
+            self._built_model = None
+            self.op_conds_to_built_models = None
+            self.op_conds_to_built_solvers = None
+
+        param = self.model.param
+        self.parameter_set = (
+            self._unprocessed_parameter_set.set_initial_stoichiometries(
+                init_soc, param=param, inplace=False
+            )
+        )
+        # Save solved initial SOC in case we need to re-build the model
+        self._built_initial_soc = init_soc
+
+    def set_params(self):
+        """
+        Set the parameters in the model.
+        """
+        if self._model_with_set_params:
+            return
+
+        self._model_with_set_params = self.parameter_set.process_model(
+            self._unprocessed_model, inplace=False
         )
+        self.parameter_set.process_geometry(self._geometry)
+        self.model = self._model_with_set_params
+
+    def step(self, signal, x, grad):
+        for i, p in enumerate(self.fit_dict):
+            self.fit_dict[p] = x[i]
+        print(self.fit_dict)
+
+        y_hat = self.solver.solve(
+            self._built_model, self.time_data, inputs=self.fit_dict
+        )[signal].data
+
+        try:
+            res = np.sqrt(
+                np.mean((self.observations["Voltage [V]"].data[1] - y_hat) ** 2)
+            )
+        except:
+            raise ValueError(
```

Comment:
I would directly check the length if you are going to raise this particular error, a ValueError could be a lot of other stuff too